WIP: A new steps/data/reverberate_data_dir.py script

tomkocse opened the pull request.

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,299 @@
+
+def GetArgs():
+    # we add compulsary arguments as named arguments for readability
+    parser = argparse.ArgumentParser(description="Generate corrupted data"
```
"Reverberate the data directory with an option to add isotropic and point source noises"

vijayaditya replied:
In the help message say that this script only deals with single channel wave files. If multi-channel noise/rir/speech files are provided one of the channels will be randomly picked.

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,299 @@
+                                     formatter_class=argparse.ArgumentDefaultsHelpFormatter)
+
+    parser.add_argument("--rir-list-file", type=str, required = True, 
+                        help="RIR information file")
```
describe the format of this file.

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,299 @@
+                        help="Number of replicate to generated for the data")
+    parser.add_argument('--snrs', type=str, dest = "snr_string", default = '20:10:0', help='snrs to be used for corruption')
+    parser.add_argument('--prefix', type=str, default = None, help='prefix for the id of the corrupted utterances')
+    parser.add_argument("--speech-rvb-probability", type=float, default = 0.8,
```
why is this needed ?

tomkocse replied:
From our previous discussion, the logic of adding rir / noises is: 
"With probability p1, choose to pick an RIR randomly.
With probability p2, choose to pick an RIR randomly from the same room or any RIR randomly if you did not pick an RIR in step 1."
So i would like to let the user to decide the probability.

vijayaditya replied:
Hmm.. ok. It doesnt hurt to keep the option.

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,299 @@
+    parser.add_argument('--prefix', type=str, default = None, help='prefix for the id of the corrupted utterances')
+    parser.add_argument("--speech-rvb-probability", type=float, default = 0.8,
+                        help="Probability of reverberating the speech signal, e.g. 0 <= p <= 1")
+    parser.add_argument("--noise-adding-probability", type=float, default = 0.4,
```
once again why is this needed ?

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,299 @@
+                        help="Probability of reverberating the speech signal, e.g. 0 <= p <= 1")
+    parser.add_argument("--noise-adding-probability", type=float, default = 0.4,
+                        help="Probability of adding point-source noises, e.g. 0 <= p <= 1")
+    parser.add_argument("--max-noises-added", type=int, default = 2,
```
why is this option needed. This should depend on the size of the recording and what if the corpora has recordings of different sizes. A better thing to control would be `--max-overlapping-noises`

tomkocse replied:
This option will be to control maximum number of point-source noises that can be added to one recording

vijayaditya replied:
yeah I understand that. I am saying the maximum number of overlapping point source noises not the number of noises that can be added to the recording.

vijayaditya replied:
Actually if you want to specify the maximum number of noises added to a file you should do something like max-number-of-noises-per-minute. You would scale this value based on the duration of the recording.

vijayaditya replied:
@tomkocse could you address the above comment.

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,299 @@
+
+    ## Check arguments.
+    if args.rir_list_file is None:
+        raise Exception("Rir information file must be provided")
```
Rir --> RIR

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,299 @@
+        os.makedirs(args.output_dir)
+
+    ## Check arguments.
+    if args.rir_list_file is None:
```
If rir_list_file is necessary you can make it a necessary argument by adding the option `required=True` while adding the option.

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,299 @@
+# The noise list would have the following format:
+# --noise-id <string,compulsary> --noise-type <choices = (isotropic, point source),compulsary> --bg-fg-type <choices=(background|foreground), default=background> --rir-file <str, compulsary if isotropic, should not be specified if point-source> < location=(support Kaldi IO strings) >
+def CorruptWav(wav_scp, durations, output_dir, room_list, noise_list, snr_string, num_replica, prefix, speech_rvb_probability, noise_adding_probability, max_noises_added):
+    rooms = list_cyclic_iterator(room_list, random_seed = 1)
```
let the user specify the random seed

danpovey replied:
Vijay, I'm not sure if Python, like Perl, sets the random seed from the
current time, but if so we should probably by default set the seed to a
constant in any Python program that uses random numbers, for
reproducibility.  We already do this in perl scripts that use random
numbers.
Dan


On Wed, Apr 20, 2016 at 12:17 PM, Vijayaditya Peddinti <
notifications@github.com> wrote:

> In egs/wsj/s5/steps/data/reverberate_data_dir.py
> <https://github.com/kaldi-asr/kaldi/pull/706#discussion_r60470528>:
>
> > +    for line in open(file, 'r'):
> > +        parts = line.split()
> > +        if assert2fields:
> > +            assert(len(parts) == 2)
> > +
> > +        dict[parts[0]] = value_processor(parts[1:])
> > +    return dict
> > +
> > +
> > +# This is the major function to generate pipeline command for the corruption
> > +# The rir list would have the following format:
> > +# --rir-id <string,compulsary> --room-id <string,compulsary> --receiver-position-id <string,optional> --source-position-id <string,optional> --rt-60 < <float,optional> --drr <float, optional> < location(support Kaldi IO strings) >
> > +# The noise list would have the following format:
> > +# --noise-id <string,compulsary> --noise-type <choices = (isotropic, point source),compulsary> --bg-fg-type <choices=(background|foreground), default=background> --rir-file <str, compulsary if isotropic, should not be specified if point-source> < location=(support Kaldi IO strings) >
> > +def CorruptWav(wav_scp, durations, output_dir, room_list, noise_list, snr_string, num_replica, prefix, speech_rvb_probability, noise_adding_probability, max_noises_added):
> > +    rooms = list_cyclic_iterator(room_list, random_seed = 1)
>
> let the user specify the random seed
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/706/files/356fa8258d848e3b092c5b894f6a7b4cab63b29c#r60470528>
>

vijayaditya replied:
I have been using a consistent default value in the python scripts I wrote, but I haven't done a check of all the scripts. I will create an issue requesting for a review of existing python scripts. We would also have to take care of the inline python scripts. 

It seems to depend on the OS.

random.seed([x])
Initialize the basic random number generator. Optional argument x can be any hashable object. If x is omitted or None, current system time is used;** current system time is also used to initialize the generator when the module is first imported. If randomness sources are provided by the operating system, they are used instead of the system time (see the os.urandom() function for details on availability).** [Reference](https://docs.python.org/2/library/random.html)

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,299 @@
+        noises = list_cyclic_iterator(noise_list, random_seed = 1)
+    snrs = list_cyclic_iterator(snr_string.split(':'))
+    command_list = []
+    for i in range(num_replica):
```
I think it would be better to assign to each RIR a probability with which it would be picked up. I did this in the ASpIRE recipe. This is necessary as we might want to boost the probability of choosing RIRs which are more relevant to the current LVCSR task. 

So in the RIR list file expect the specification of a probability for each RIR, if nothing make it the 1/num_rirs.
Finally after reading in the entire list and assigning probabilities to RIRs which have none assigned, renormalize to ensure the RIR probability distribution sums to one.

This probability vector to sample from the RIR list accordingly. IIRC code for doing this is there in the ASpIRE recipe.

@vimalmanohar you should probability also assign a probability to the noise signals. Not all noises are equally important for all recipes.

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,299 @@
+        for wav_id in keys:
+            wav_pipe = wav_scp[wav_id]
+            wav_dur = durations[wav_id]
+            if prefix is not None:
```
You are always specifying a prefix so the check is not necessary.

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,299 @@
+# --rir-id <string,compulsary> --room-id <string,compulsary> --receiver-position-id <string,optional> --source-position-id <string,optional> --rt-60 < <float,optional> --drr <float, optional> < location(support Kaldi IO strings) >
+# The noise list would have the following format:
+# --noise-id <string,compulsary> --noise-type <choices = (isotropic, point source),compulsary> --bg-fg-type <choices=(background|foreground), default=background> --rir-file <str, compulsary if isotropic, should not be specified if point-source> < location=(support Kaldi IO strings) >
+def CorruptWav(wav_scp, durations, output_dir, room_list, noise_list, snr_string, num_replica, prefix, speech_rvb_probability, noise_adding_probability, max_noises_added):
```
process the snr_string before calling this function. Better provide SNRs as a float array.

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,299 @@
+    wav_scp = ParseFileToDict(input_dir + "/wav.scp", value_processor = lambda x: " ".join(x))
+    CorruptWav(wav_scp, durations, output_dir, room_list, noise_list, snr_string, num_replica, prefix, speech_rvb_probability, noise_adding_probability, max_noises_added)
+
+    ReplicateFileType2(input_dir + "/utt2spk", output_dir + "/utt2spk", num_replica, prefix)
```
Choose a different name for this function

vijayaditya replied:
Please choose a different name for this function.

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,299 @@
+    f.close()
+
+
+def MakeCorruption(input_dir, output_dir, room_list, noise_list, snr_string, num_replica, prefix, speech_rvb_probability, noise_adding_probability, max_noises_added):
```
Rename this function as CreateReverberatedCopy

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,299 @@
+
+
+# This function replicate the entries in files like text
+def ReplicateFileType1(input_file, output_file, num_replica, prefix):
```
Two functions are not necessary. Write a single function 
`AddPrefixToFields(prefix, input_file, output_file, num_replica, fields = [0])`

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,299 @@
+    for i in range(num_replica):
+        for line in list:
+            split1 = line.split()
+            if prefix is not None:
```
no need for this check as prefix is a compulsary argument.

tomkocse replied:
prefix is not compulsory when --num-replications is 1,  better not to add prefix for test set like eval2000 as it need to associated with the glm/stm file

vimalmanohar replied:
I think it is better do add prefix at all times. You will never by reverberating test sets and even if you do, you should add the prefix to appropriate columns in both reco2file_and_channel and stm. GLM should not change because it has nothing to do with the file ids.

tomkocse replied:
We are currently doing some experiment that need to reverberate the test sets (e.g. eval2000), the local/score_sclite.sh need to be modified accordingly even we add prefix to appropriate columns  of "stm" file

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,299 @@
+                for j in range(len(rir_list)):
+                    if noise.rir_file == rir_list[j].rir_file_location:
+                        id = j
+                        print(noise.rir_file)
```
remove this print statement.

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,299 @@
+                    noise_rir = room['rir_list'][random.randint(0,len(room['rir_list'])-1)]
+                    start_times_added.append(round(random.random() * wav_dur, 2))
+                    noises_added.append("\"wav-reverberate --duration={2} --impulse-response={1} {0} - |\" ".format(noise.noise_file_location, noise_rir.rir_file_location, round(random.random()*(wav_dur-start_times_added[-1]), 2)))
+                    snrs_added.append("{0}".format(snrs.next()))
```
@vimalmanohar could you also please review this PR.

vimalmanohar replied:
Make the snr dependent on foreground vs background noise. Have two parameters input from the user --background-snrs and --foreground-snrs and then depending on which type of noise is drawn, you can do either choose snr from foreground_snrs or background_snrs.

tomkocse replied:
@vimalmanohar So will isotropic noise always  be treated as background noise and use the snrs from background_snrs? And will isotropic noise always be extended and cover the whole file of the speech waveform ?

vijayaditya replied:
Yes.

On Thu, Apr 21, 2016 at 4:44 AM, tomkocse <notifications@github.com> wrote:

> In egs/wsj/s5/steps/data/reverberate_data_dir.py
> <https://github.com/kaldi-asr/kaldi/pull/706#discussion_r60544896>:
>
> > +                command_opts += "--impulse-response={0} ".format(speech_rir.rir_file_location)
> > +                # add the corresponding isotropic noise if there is any
> > +                if len(speech_rir.iso_noise_list) > 0:
> > +                    isotropic_noise = speech_rir.iso_noise_list[random.randint(0,len(speech_rir.iso_noise_list)-1)]
> > +                    noises_added.append("{0}".format(isotropic_noise.noise_file_location))
> > +                    snrs_added.append("{0}".format(snrs.next()))
> > +                    start_times_added.append(round(random.random() * wav_dur, 2))
> > +
> > +            if noises is not None and random.random() < noise_adding_probability:
> > +                for k in range(random.randint(1, max_noises_added)):
> > +                    # pick the RIR to reverberate the point-source noise
> > +                    noise = noises.next()
> > +                    noise_rir = room['rir_list'][random.randint(0,len(room['rir_list'])-1)]
> > +                    start_times_added.append(round(random.random() * wav_dur, 2))
> > +                    noises_added.append("\"wav-reverberate --duration={2} --impulse-response={1} {0} - |\" ".format(noise.noise_file_location, noise_rir.rir_file_location, round(random.random()*(wav_dur-start_times_added[-1]), 2)))
> > +                    snrs_added.append("{0}".format(snrs.next()))
>
> @vimalmanohar <https://github.com/vimalmanohar> So will isotropic noise
> always be treated as background noise and use the snrs from
> background_snrs? And will isotropic noise always be extended and cover the
> whole file of the speech waveform ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/706/files/356fa8258d848e3b092c5b894f6a7b4cab63b29c#r60544896>
>

vimalmanohar commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,299 @@
+                # add the corresponding isotropic noise if there is any
+                if len(speech_rir.iso_noise_list) > 0:
+                    isotropic_noise = speech_rir.iso_noise_list[random.randint(0,len(speech_rir.iso_noise_list)-1)]
+                    noises_added.append("{0}".format(isotropic_noise.noise_file_location))
```
Not a very critical thing -- but you can probably use `noises_added.append(isotropic_noise.noise_file_location)`

vimalmanohar commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,299 @@
+                if len(speech_rir.iso_noise_list) > 0:
+                    isotropic_noise = speech_rir.iso_noise_list[random.randint(0,len(speech_rir.iso_noise_list)-1)]
+                    noises_added.append("{0}".format(isotropic_noise.noise_file_location))
+                    snrs_added.append("{0}".format(snrs.next()))
```
As Vijay had mentioned, you can keep the snrs as float values and then use `snrs_added.append(snrs.next())`

vimalmanohar replied:
As mentioned below, make this background_snrs.next() as this can probably be treated as background noise.

vimalmanohar commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,299 @@
+                    noises_added.append("\"wav-reverberate --duration={2} --impulse-response={1} {0} - |\" ".format(noise.noise_file_location, noise_rir.rir_file_location, round(random.random()*(wav_dur-start_times_added[-1]), 2)))
+                    snrs_added.append("{0}".format(snrs.next()))
+
+            if len(noises_added) > 1:
```
This should be > 0, I think.

vimalmanohar commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,299 @@
+
+            if len(noises_added) > 1:
+                command_opts += "--additive-signals='{0}' ".format(','.join(noises_added))
+            if len(snrs_added) > 1:
```
Same here and the next line

vimalmanohar commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,299 @@
+            if command_opts == "":
+                command = "{0} {1}\n".format(wav_id, wav_pipe) 
+            else:
+                command = "{0} {1} wav-reverberate {2} - - |\n".format(wav_id, wav_pipe, command_opts)
```
You should check somewhere that this is indeed a pipe. You can have just a file name in some cases. In that case you might have to do `wav_pipe = 'cat {wav_file} |'`

vimalmanohar commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,299 @@
+            command_opts = ""
+            noises_added = []
+            snrs_added = []
+            start_times_added = []
```
I think there should be another for durations of noise added. For isotropic, you can make it the whole file. For others, you have to choose randomly. We could probably have choose some random duration < max-duration. It might be better to make it separate paramters for background and foregroud noises because the foreground noises are short and you don't want to make the duration too long. 
@vijayaditya Do you have any suggestions regarding how to choose duration?

vijayaditya replied:
@vimalmanohar Could you give examples of the point source noises you have access to. If it is things like door bangs I don't think we should be repeating them. If it is things like keyboard sounds  it makes sense to repeat them. I am not aware of the complete list of point source noises you plan to use, it might be better to decide based on that.

For the sake of ensuring all of us are on the same page could you give some examples which explain the critical differences between these noises in your mind. It would be better if you could give some examples where people might have confusion classifying the noises.

vimalmanohar replied:
Background noise is like stationary noise or like stationary noise. It is
long and has the same characteristics for a long duration. It can be noise
like background AC noise or even like a siren or bird sound or perhaps even
music. The foreground noise would be something of short duration like door
taps, keyboard sound, bells, footsteps, dog bark, mic movements etc.
Duration would be the critical difference. Or to be more specific, the
noise energy is concentrated over a short duration for a "foreground"
noise. The foreground noise files I have are not necessarily short files.
They may have the foreground noise repeated after a period of silence (e.g.
door taps separated by 1s of silence). It should ideally not have such
silences, but it might not be possible to get well conditioned noises.


On Thu, Apr 21, 2016 at 3:34 PM Vijayaditya Peddinti <
notifications@github.com> wrote:

> In egs/wsj/s5/steps/data/reverberate_data_dir.py
> <https://github.com/kaldi-asr/kaldi/pull/706#discussion_r60642044>:
>
> > +    for i in range(num_replica):
> > +        keys = wav_scp.keys()
> > +        keys.sort()
> > +        for wav_id in keys:
> > +            wav_pipe = wav_scp[wav_id]
> > +            wav_dur = durations[wav_id]
> > +            if prefix is not None:
> > +                wav_id = prefix + str(i) + "_" + wav_id
> > +            command = "{0} {1} wav-reverberate".format(wav_id, wav_pipe)
> > +
> > +            # pick the room
> > +            room = rooms.next()
> > +            command_opts = ""
> > +            noises_added = []
> > +            snrs_added = []
> > +            start_times_added = []
>
> @vimalmanohar <https://github.com/vimalmanohar> Could you give examples
> of the point source noises you have access to. If it is things like door
> bangs I don't think we should be repeating them. If it is things like
> keyboard sounds it makes sense to repeat them. I am not aware of the
> complete list of point source noises you plan to use, it might be better to
> decide based on that.
>
> For the sake of ensuring all of us are on the same page could you give
> some examples which explain the critical differences between these noises
> in your mind. It would be better if you could give some examples where
> people might have confusion classifying the noises.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/706/files/356fa8258d848e3b092c5b894f6a7b4cab63b29c#r60642044>
>
-- 
Vimal Manohar
PhD Student
Electrical & Computer Engineering
Johns Hopkins University

tomkocse commented:
In the current commit, background noise is always extended to the length of speech and added at the start of the speech; foreground noise is always not extended and added at the random point of the speech. For more detailed noise adding mechanism i guess we have to add more fields to the noise information list so that user can have a better control to the noise addition.
Sampling of RIR according to different RIR probability is not yet handled.
--max-noised-added option is not yet modified as it is difficult to control the number of overlapped noise occurred at the same time. I need to be more clear about the noise adding mechanism before i can go on.

vimalmanohar commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -156,52 +169,45 @@ def CorruptWav(wav_scp, durations, output_dir, room_list, noise_list, snr_string
-        ReplicateFileType2(input_dir + "/reco2file_and_channel", output_dir + "/reco2file_and_channel", num_replica, prefix)
+        AddPrefixToFields(input_dir + "/reco2file_and_channel", output_dir + "/reco2file_and_channel", num_replica, prefix, field = [0,1])
 
     train_lib.RunKaldiCommand("utils/validate_data_dir.sh --no-feats {output_dir}"
```
You need to add --no-text in case there is no text.

danpovey commented:
@vijayaditya, what the status of this pull request?  Is it OK as far as you're concerned?
Does it need to be committed soon or should it be marked WIP?

tomkocse commented:
I am still working on it, i just marked it as WIP

vijayaditya commented:
It needs to be combined with updates to wav-reverberate.cc so it will be in
WIP status for a significant amount of time.

--Vijay

On Sun, May 1, 2016 at 7:43 PM, Daniel Povey <notifications@github.com>
wrote:

> @vijayaditya <https://github.com/vijayaditya>, what the status of this
> pull request? Is it OK as far as you're concerned?
> Does it need to be committed soon or should it be marked WIP?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/706#issuecomment-216082013>
>

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,363 @@
+def PickItemFromDict(dict):
+   total_p = sum(dict[key].probability for key in dict.keys())
+   p = random.uniform(0, total_p)
+   upto = 0
```
change the variable name `upto`

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,363 @@
+    return args
+
+
+def PickItemFromDict(dict):
```
You don't need two methods `PickItemFromDict` and `PickItemFromList`. You could check the type of the input argument inside the method and convert to list if the input is a dict.

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,363 @@
+      if upto + item.probability >= p:
+         return item
+      upto += item.probability
+   assert False, "Shouldn't get here"
```
The message should be more informative.

vijayaditya commented:
@tomkocse After going through some of the code I am wondering if you tested the python script. Could you please create some sample RIR and noise lists and check the output of the script.

tomkocse commented:
@vijayaditya I have definitely tested the python script with my sample RIR and noise lists. One possible reason is that my sample noise lists are not complicated enough to test all the cases. It will be good if @vimalmanohar  can give me his noise lists. 

Or could you please tell me any part of the code you feel not comfortable ?

vijayaditya commented:
@tomkocse OK let me send you some possible cases for RIR lists and noise lists.
@vimalmanohar Please add what you think is necessary to be tested.

vijayaditya commented:
@tomkocse Could you add the bug-fix discussed in #889 . I am performing a review of this PR now.

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,363 @@
+                   noise_list = noise_list,
+                   foreground_snr_string = args.foreground_snr_string,
+                   background_snr_string = args.background_snr_string,
+                   num_replica = args.num_replica,
```
`num_replicas`

tomkocse commented:
@vijayaditya As you suggested, an option --shift-output will be added to wav-reverberate.cc , when it is true, the output will be shift and the output signal length will be equal to the input signal length; when it is false, then output signal length = input signal length + rir length -1

vijayaditya commented:
Remember to make it true by default.

--Vijay

On Mon, Jul 11, 2016 at 1:39 AM, tomkocse <notifications@github.com> wrote:

> @vijayaditya <https://github.com/vijayaditya> As you suggested, an option
> --shift-output will be added to wav-reverberate.cc , when it is true, the
> output will be shift and the output signal length will be equal to the
> input signal length; when it is false, then output signal length = input
> signal length + rir length -1
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/706#issuecomment-231646384>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/ADtwoGb7o_TJOaV3VE05M0FqisgBbO-1ks5qUdcCgaJpZM4IJM2C>
> .
>

vijayaditya commented:
@tomkocse Could you please address some of the other comments I made in the previous review, I will continue reviewing after that.

tomkocse commented:
@vijayaditya Previous comments addressed

tomkocse commented:
@vijayaditya Please let me know if there are other comments needed to be addressed

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,352 @@
+from __future__ import print_function
+import argparse, glob, math, os, random, sys, warnings, copy, imp, ast
+
+train_lib = imp.load_source('ntl', 'steps/nnet3/nnet3_train_lib.py')
```
Including the nnet3 library just for `RunKaldiCommand` is not desirable. For now just write `steps/data/data_lib.py` where the `RunKaldiCommand` is replicated. We can think of a better way to share `RunKaldiCommand` later.

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,352 @@
+   total_p = sum(item.probability for item in plist)
+   p = random.uniform(0, total_p)
+   accumulate_p = 0
+   for item in plist:
```
Are you assuming that the list of probability values is sorted in some order ?

tomkocse replied:
No, they don't have to be sorted

vijayaditya replied:
Specify that you are using the inverse transform sampling method.

vijayaditya replied:
This sampling will be erroneous when x is a dictionary, as each time x.values() is called the returned values can be ordered differently. If you do not maintain the ordering of the items in your collection the inverse sampling method will not work.

danpovey replied:
I don't see that it matters, I think it will sample correctly.

vijayaditya replied:
As the ordering of sets is not maintained across calls the inverse cumulative distribution function would change in each call.

danpovey replied:
It's only calling x.values() once within that function, not more than once.

The only scenario there might be a problem is where an adversary with
knowledge of the random number generator is reordering the values in a
special way each time.  If the random numbers are truly random this
wouldn't be even possible.  Even if it were possible, it's paranoid.


On Tue, Jul 19, 2016 at 1:25 PM, Vijayaditya Peddinti <
notifications@github.com> wrote:

> In egs/wsj/s5/steps/data/reverberate_data_dir.py
> <https://github.com/kaldi-asr/kaldi/pull/706#discussion_r71416268>:
>
> > +        warnings.warn("--prefix is set to 'rvb' as --num-replications is larger than 1.")
> > +
> > +    return args
> > +
> > +
> > +# This function pick the item according to the associated probability
> > +# The input could be either a dictinoary of a list
> > +def PickItemWithProbability(x):
> > +   if isinstance(x, dict):
> > +     plist = list(set(x.values()))
> > +   else:
> > +     plist = x
> > +   total_p = sum(item.probability for item in plist)
> > +   p = random.uniform(0, total_p)
> > +   accumulate_p = 0
> > +   for item in plist:
>
> As the ordering of sets is not maintained across calls the inverse
> cumulative distribution function would change in each call.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/706/files/3802fde4b4da335c1482da6519f3f8d58145200f#r71416268>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADJVu2lpbSPHxbQ9Tdejs7Hko2IEevkFks5qXTKxgaJpZM4IJM2C>
> .
>

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,352 @@
+   assert False, "Shouldn't get here as the accumulated probability should always equal to 1"
+
+
+def ParseFileToDict(file, assert2fields = False, value_processor = None):
```
Have one argument per line and describe them in detail.

For every method describe in a comment the purpose of the method.

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,352 @@
+    return dict
+
+
+# This is the major function to generate pipeline command for the corruption
```
`This is the main function`

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,352 @@
+# The generic command of wav-reverberate will be like:
+# wav-reverberate --duration=t --impulse-response=rir.wav 
+# --additive-signals='noise1.wav,noise2.wav' --snrs='snr1,snr2' --start-times='s1,s2' input.wav output.wav
+def CorruptWav(wav_scp, durations, output_dir, room_dict, noise_list, foreground_snr_array, background_snr_array, num_replicas, prefix, speech_rvb_probability, noise_adding_probability, max_noises_per_minute):
```
One argument per line please, along with a description of the argument wherever necessary.

vijayaditya replied:
describe the format of `room_dict`.

vijayaditya replied:
This function is beyond the maximum line limit. I would **strongly** recommend breaking it into smaller functions.

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,352 @@
+                    start_times_added.append(0)
+
+            # Add the point-source noise
+            if len(noise_list) > 0 and random.random() < noise_adding_probability:
```
Move the definition of `max_noises_per_recording` here.

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,352 @@
+    f.close()
+
+
+def CreateReverberatedCopy(input_dir, output_dir, room_dict, noise_list, foreground_snr_string, background_snr_string, num_replicas, prefix, speech_rvb_probability, noise_adding_probability, max_noises_per_minute):
```
Add a comment describing this function. Have one line per argument and describe each argument (and its format) in detail.

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,352 @@
+            item.probability = uniform_probability
+        else:
+            # smooth the probability
+            item.probability = 0.3 * item.probability + 0.7 * uniform_probability
```
why is uniform distribution getting such a large weight ?

tomkocse replied:
@vijayaditya I remember we both agree on using 0.7 for the weight of the uniform_probability. This value decides how heavy you want the smoothing to be

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,352 @@
+    train_lib.RunKaldiCommand("utils/validate_data_dir.sh --no-feats {output_dir}"
+                    .format(output_dir = output_dir))
+
+def SmoothProbability(list):
```
Add a comment describing the goal of this function. Describe the format of list.
Rename this function to `SmoothProbabilityDistribution`.

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,352 @@
+            # smooth the probability
+            item.probability = 0.3 * item.probability + 0.7 * uniform_probability
+
+    sum_p = sum(item.probability for item in list)
```
`sum_p` can be moved to previous for loop.

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,352 @@
+    return SmoothProbability(rir_list)
+
+
+# This function crate the room dictinoary from the rir list
```
```This functions creates the room dictionary, indexed by the room id, from the rir list.```

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,352 @@
+
+# This function crate the room dictinoary from the rir list
+# The key of the returned dictionary is the room id
+def MakeRoomDict(rir_list):
```
describe the format of the `rir_list`.

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,352 @@
+    for key in room_dict.keys():
+        room_dict[key].probability = sum(rir.probability for rir in room_dict[key].rir_list)
+
+    return room_dict
```
describe the format of `room_dict`.

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,352 @@
+            if len(noise_list) > 0 and random.random() < noise_adding_probability:
+                for k in range(random.randint(1, max_noises_recording)):
+                    # pick the RIR to reverberate the point-source noise
+                    noise = PickItemWithProbability(noise_list)
```
I am assuming that this scripts allows for simple addition of background noises without reverberation. Is this correct ?

tomkocse replied:
@vijayaditya for point-source noise, the noise itself will always be reverberated by its associated RIR before adding to the speech wave. For isotropic noise, the addition is independent with the reverberation of the speech wave, i.e. one can set the rvb probability to zero so that the script will always add noises without reverberation of the speech. I will explain more in my next commit

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,352 @@
+                        noises_added.append("wav-reverberate --impulse-response={1} {0} - |".format(noise.noise_file_location, noise_rir.rir_file_location))
+                        snrs_added.append(foreground_snrs.next())
+
+            if len(noises_added) > 0:
```
Shouldn't the length of these three lists be the same ? Assert this.

vijayaditya commented on src/feat/signal.cc:
```diff
@@ -34,31 +34,34 @@ void ElementwiseProductOfFft(const Vector<BaseFloat> &a, Vector<BaseFloat> *b) {
   int32 filter_length = filter.Dim();
-  Vector<BaseFloat> signal_padded(signal_length + filter_length - 1);
+  int32 output_length = signal_length + filter_length - 1;
+  Vector<float> signal_padded(output_length);
```
Is there a specific reason to `float` type in place of `BaseFloat` type ?

tomkocse replied:
@vijayaditya Um.... no specific reason, what is the difference of using float instead of BaseFloat ?

vijayaditya replied:
Using `BaseFloat` allows for the control of precision using `KALDI_DOUBLE_PRECISION` variable. I would strongly recommend following this precision specification and using `BaseFloat`.

vijayaditya commented on src/featbin/wav-reverberate.cc:
```diff
@@ -107,23 +137,36 @@ int main(int argc, char *argv[]) {
+                "the output waveform will be equal to the input waveform."
+                "If false, the length of the output waveform will be "
+                "equal to (original input length + rir length - 1). "
+                "This value is default true and "
```
`This value is true by default.`

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,352 @@
+    return item
+
+
+def GetArgs():
```
make `GetArgs()` the first method definition in the file.

vijayaditya commented:
Completed one round of review.

tomkocse commented:
@vijayaditya Do you mean posting the output wav.scp ?

vijayaditya commented:
Yes post a wav.scp which has all the examples of interest.

--Vijay

On Fri, Jul 15, 2016 at 12:16 PM, tomkocse <notifications@github.com> wrote:

> @vijayaditya <https://github.com/vijayaditya> Do you mean posting the
> output wav.scp ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/706#issuecomment-232997624>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADtwoFk8WS93J4NaXViX7o1Ry6VNHgaQks5qV7J5gaJpZM4IJM2C>
> .
>

vijayaditya commented:
@tomkocse Did you verify that the reverberated and original signals are aligned, by default ?

vijayaditya commented on egs/wsj/s5/steps/data/data_lib.py:
```diff
@@ -0,0 +1,23 @@
+import subprocess
```
You could follow @jtrmal 's suggestion and rename this file. May be something like `data_dir_manipulation_lib.py`. 

Remove the libraries which have been commented out.

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,470 @@
+from __future__ import print_function
+import argparse, glob, math, os, random, sys, warnings, copy, imp, ast
+
+data_lib = imp.load_source('ntl', 'steps/data/data_lib.py')
```
Name `ntl` does not seem appropriate for the imported library.

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,470 @@
+def GetArgs():
+    # we add compulsary arguments as named arguments for readability
+    parser = argparse.ArgumentParser(description="Reverberate the data directory with an option "
+                                                 "to add isotropic and point source noiseis. "
```
`noiseis`-->`noises`.

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,470 @@
+                        help="RIR information file, the format of the file is "
+                        "--rir-id <string,compulsary> --room-id <string,compulsary> "
+                        "--receiver-position-id <string,optional> --source-position-id <string,optional> "
+                        "--rt-60 < <float,optional> --drr <float, optional> < location(support Kaldi IO strings) >")
```
```python
--rt-60 <float,optional> --drr <float, optional> < location(support Kaldi IO strings)>
```

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,470 @@
+    parser.add_argument("--rir-list-file", type=str, required = True, 
+                        help="RIR information file, the format of the file is "
+                        "--rir-id <string,compulsary> --room-id <string,compulsary> "
+                        "--receiver-position-id <string,optional> --source-position-id <string,optional> "
```
You could provide an example.

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,470 @@
+                        "--rt-60 < <float,optional> --drr <float, optional> < location(support Kaldi IO strings) >")
+    parser.add_argument("--noise-list-file", type=str, default = None,
+                        help="Noise information file, the format of the file is"
+                        "--noise-id <string,compulsary> --noise-type <choices = (isotropic, point source),compulsary> "
```
It seems appropriate to enclose the choices in "{}" rather than "()".

tomkocse replied:
I saw that in your other script you use "[]" for choices , so which one you prefer?

danpovey replied:
btw, it's spelled "compulsory", but I would use the word "required" instead.
Dan


On Tue, Jul 19, 2016 at 8:14 PM, tomkocse <notifications@github.com> wrote:

> In egs/wsj/s5/steps/data/reverberate_data_dir.py
> <https://github.com/kaldi-asr/kaldi/pull/706#discussion_r71459791>:
>
> > +                                                 "to add isotropic and point source noiseis. "
> > +                                                 "Usage: reverberate_data_dir.py [options...] <in-data-dir> <out-data-dir> "
> > +                                                 "E.g. reverberate_data_dir.py --rir-list-file rir_list "
> > +                                                 "--foreground-snrs 20:10:15:5:0 --background-snrs 20:10:15:5:0 "
> > +                                                 "--noise-list-file noise_list --speech-rvb-probability 1 --num-replications 2 "
> > +                                                 "--random-seed 1 data/train data/train_rvb",
> > +                                     formatter_class=argparse.ArgumentDefaultsHelpFormatter)
> > +
> > +    parser.add_argument("--rir-list-file", type=str, required = True,
> > +                        help="RIR information file, the format of the file is "
> > +                        "--rir-id <string,compulsary> --room-id <string,compulsary> "
> > +                        "--receiver-position-id <string,optional> --source-position-id <string,optional> "
> > +                        "--rt-60 < <float,optional> --drr <float, optional> < location(support Kaldi IO strings) >")
> > +    parser.add_argument("--noise-list-file", type=str, default = None,
> > +                        help="Noise information file, the format of the file is"
> > +                        "--noise-id <string,compulsary> --noise-type <choices = (isotropic, point source),compulsary> "
>
> I saw that in your other script you use "[]" for choices , so which one
> you prefer?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/706/files/d335c718e3b1849afe40a61fc436c0e098c6a36e#r71459791>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADJVu54zkMntG4E2J0t9USNTsuY-JQygks5qXZKwgaJpZM4IJM2C>
> .
>

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,470 @@
+                        help="Noise information file, the format of the file is"
+                        "--noise-id <string,compulsary> --noise-type <choices = (isotropic, point source),compulsary> "
+                        "--bg-fg-type <choices=(background|foreground), default=background> "
+                        "--rir-file <str, compulsary if isotropic, should not be specified if point-source> "
```
```---rir-file <str, specifies the rir file associated with the noise file. Compulsory if isotropic as the rir file links this noise file to a specific position in the room.```

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,470 @@
+                        "< location=(support Kaldi IO strings) >")
+    parser.add_argument("--num-replications", type=int, dest = "num_replicas", default = 1,
+                        help="Number of replicate to generated for the data")
+    parser.add_argument('--foreground-snrs', type=str, dest = "foreground_snr_string", default = '20:10:0', help='snrs for foreground noises')
```
`When foreground noises are being added the script will iterate through these SNRs.`

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,470 @@
+    parser.add_argument("--num-replications", type=int, dest = "num_replicas", default = 1,
+                        help="Number of replicate to generated for the data")
+    parser.add_argument('--foreground-snrs', type=str, dest = "foreground_snr_string", default = '20:10:0', help='snrs for foreground noises')
+    parser.add_argument('--background-snrs', type=str, dest = "background_snr_string", default = '20:10:0', help='snrs for background noises')
```
`When background noises are being added the script will iterate through these SNRs.`

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,470 @@
+                        help="Number of replicate to generated for the data")
+    parser.add_argument('--foreground-snrs', type=str, dest = "foreground_snr_string", default = '20:10:0', help='snrs for foreground noises')
+    parser.add_argument('--background-snrs', type=str, dest = "background_snr_string", default = '20:10:0', help='snrs for background noises')
+    parser.add_argument('--prefix', type=str, default = None, help='prefix for the id of the corrupted utterances')
```
`This prefix will modified for each reverberated copy, by adding additional affixes.`

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,470 @@
+                        help="Probability of adding isotropic noises, e.g. 0 <= p <= 1")
+    parser.add_argument("--max-noises-per-minute", type=int, default = 2,
+                        help="This controls the maximum number of point-source noises that could be added to a recording according to its duration")
+    parser.add_argument('--random-seed', type=int, default=0, help='seed to be used in the randomization of impulese and noises')
```
`impulese`-->`impuleses`

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,470 @@
+            if type(value) is tuple:
+                value = list(value)
+            value.sort()
+            value = ' '.join(value)
```
convert the items in the collection `value` to strings.

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,470 @@
+            # pick the RIR to reverberate the point-source noise
+            noise = PickItemWithProbability(point_noise_list)
+            noise_rir = PickItemWithProbability(room.rir_list)
+            if noise.bg_fg_type == "background":
```
Clearly mention how background and foreground noises are treated differently here.

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,470 @@
+                                 'start_times': [],
+                                 'snrs': []}
+    # Randomly select the room
+    room = PickItemWithProbability(room_dict)
```
It would be better to mention here that the room probability is a sum of the probabilities of the RIR recorded in the room.

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,470 @@
+        # pick the RIR to reverberate the speech
+        reverberate_opts += "--impulse-response={0} ".format(speech_rir.rir_file_location)
+
+    rir_iso_noise_list = FilterIsotropicNoiseList(iso_noise_list, speech_rir.rir_id)
```
Do you want to call FilterIsotropicNoiseList for each recording ? It might be better to prepare a dictionary of istropic noises indexed by rir_id and just use it across calls to method.

tomkocse replied:
Can we leave it here unless performance degradation is found ?

vijayaditya replied:
OK, you can profile the code when building the multi-database multi-condition recipe.

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,470 @@
+
+    noise_addition_descriptor = AddPointSourceNoise(noise_addition_descriptor,  # descriptor to store the information of the noise added
+                                                    room,  # the room selected
+                                                    point_noise_list, # the point source noise list
```
Rename `point_noise_list` as `pointsource_noise_list`.

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,470 @@
+# The generic command of wav-reverberate will be like:
+# wav-reverberate --duration=t --impulse-response=rir.wav 
+# --additive-signals='noise1.wav,noise2.wav' --snrs='snr1,snr2' --start-times='s1,s2' input.wav output.wav
+def CorruptWav(wav_scp,  # the dictionary of which elements are the IO of the speech recordings
```
`CorruptWav` --> `GenerateReverberatedWavScp`

vijayaditya replied:
`#the dictionary of which elements are the IO of the speech recordings`
-->
`#a dictionary whose values are the Kaldi-IO strings of the speech recordings

danpovey replied:
The standard term for this is 'rxfilename'



On Tue, Jul 19, 2016 at 2:03 PM, Vijayaditya Peddinti <
notifications@github.com> wrote:

> In egs/wsj/s5/steps/data/reverberate_data_dir.py
> <https://github.com/kaldi-asr/kaldi/pull/706#discussion_r71422914>:
>
> > +
> > +    assert len(noise_addition_descriptor['noise_io']) == len(noise_addition_descriptor['start_times'])
> > +    assert len(noise_addition_descriptor['noise_io']) == len(noise_addition_descriptor['snrs'])
> > +    if len(noise_addition_descriptor['noise_io']) > 0:
> > +        reverberate_opts += "--additive-signals='{0}' ".format(','.join(noise_addition_descriptor['noise_io']))
> > +        reverberate_opts += "--start-times='{0}' ".format(','.join(map(lambda x:str(x), noise_addition_descriptor['start_times'])))
> > +        reverberate_opts += "--snrs='{0}' ".format(','.join(map(lambda x:str(x), noise_addition_descriptor['snrs'])))
> > +
> > +    return reverberate_opts
> > +
> > +
> > +# This is the main function to generate pipeline command for the corruption
> > +# The generic command of wav-reverberate will be like:
> > +# wav-reverberate --duration=t --impulse-response=rir.wav
> > +# --additive-signals='noise1.wav,noise2.wav' --snrs='snr1,snr2' --start-times='s1,s2' input.wav output.wav
> > +def CorruptWav(wav_scp,  # the dictionary of which elements are the IO of the speech recordings
>
> #the dictionary of which elements are the IO of the speech recordings
> -->
> `#a dictionary whose values are the Kaldi-IO strings of the speech
> recordings
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/706/files/d335c718e3b1849afe40a61fc436c0e098c6a36e#r71422914>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADJVu6n4Fgl9AK7KhWxsSSfPojQ_kJq-ks5qXTufgaJpZM4IJM2C>
> .
>

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,470 @@
+            speech_dur = durations[wav_id]
+            max_noises_recording = math.floor(max_noises_per_minute * speech_dur / 60)
+            if prefix is not None:
+                new_wav_id = prefix + str(i) + "_" + wav_id
```
Changing the naming convention would require several modifications spread across different methods. I would recommend passing around a function pointer, where you define the naming convention. You can call this function with the old ids (speaker id /utterance id/ recording id), prefix and replica number and get the new ids.

danpovey replied:
Is there a reason it can't just be a global function?  I don't like
function pointers unless there's some really good reason.



On Tue, Jul 19, 2016 at 2:10 PM, Vijayaditya Peddinti <
notifications@github.com> wrote:

> In egs/wsj/s5/steps/data/reverberate_data_dir.py
> <https://github.com/kaldi-asr/kaldi/pull/706#discussion_r71424128>:
>
> > +               ):
> > +    foreground_snrs = list_cyclic_iterator(foreground_snr_array)
> > +    background_snrs = list_cyclic_iterator(background_snr_array)
> > +    corrupted_wav_scp = {}
> > +    for i in range(num_replicas):
> > +        keys = wav_scp.keys()
> > +        keys.sort()
> > +        for wav_id in keys:
> > +            wav_original_pipe = wav_scp[wav_id]
> > +            # check if it is really a pipe
> > +            if len(wav_original_pipe.split()) == 1:
> > +                wav_original_pipe = "cat {0} |".format(wav_original_pipe)
> > +            speech_dur = durations[wav_id]
> > +            max_noises_recording = math.floor(max_noises_per_minute * speech_dur / 60)
> > +            if prefix is not None:
> > +                new_wav_id = prefix + str(i) + "_" + wav_id
>
> Changing the naming convention would require several modifications spread
> across different methods. I would recommend passing around a function
> pointer, where you define the naming convention. You can call this function
> with the old ids (speaker id /utterance id/ recording id), prefix and
> replica number and get the new ids.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/706/files/d335c718e3b1849afe40a61fc436c0e098c6a36e#r71424128>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADJVu9oAvQD-F6IcrWvsExbW3ZXJ2Egmks5qXT1CgaJpZM4IJM2C>
> .
>

vijayaditya replied:
Yeah a global function would be better, sorry I was thinking about some other PR when suggesting function pointers.

tomkocse replied:
@vijayaditya Can you explicitly tell me how you want to change this ?

vijayaditya replied:
Have a function like this

```python
def  GetNewId(old_utt_id, old_spk_id, old_wav_id, prefix=None, copy=0):
   .................
   .................
   return [new_utt_id, new_spk_id, new_wav_id]

```

tomkocse replied:
@vijayaditya Sorry i don't get what you mean. What i need here is just the recording id, why do i need to inpu the spk_id ?

vijayaditya replied:
I am suggesting that you move all the code which decides the new ids to a single function or multiple functions, whatever suits your purpose. As this  new id generation code is isolated from the rest of reverberation code it would be easy to modify and keep track of.

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,470 @@
+    for rir in rir_list:
+        if rir.room_id not in room_dict:
+            # add new room
+            room_dict[rir.room_id] = lambda: None
```
Why is the value a lambda function?

tomkocse replied:
This is just a way to new a object, i will modify the comment

tomkocse commented:
@vijayaditya Yes i have compared the waveform of the signals, the shifting bug has been fixed, and by default --shift-output is true

vijayaditya commented:
@tomkocse Would this be ready for another review today ? Could you prioritize this PR ?

tomkocse commented:
@vijayaditya Yes, i have tried to address the suggested comments as much as possible

vijayaditya commented:
@tomkocse  Could you please update one of the recipes which uses data reverberation to use this new script. I think ASpIRE is the only recipe in master branch which uses reverberation, so you can update that.

(For now prepare  a basic RIR list and noise list using the information available in `data/impulses_noises/`)

tomkocse commented:
@vijayaditya Do you mean making the rir_list and noise_list for the real RIR and noises in the Aspire recipe ?

vijayaditya commented:
Yes.

Vijay

On Jul 22, 2016 8:57 PM, "tomkocse" <notifications@github.com> wrote:

> @vijayaditya <https://github.com/vijayaditya> Do you mean making the
> rir_list and noise_list for the real RIR and noises in the Aspire recipe ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/706#issuecomment-234689389>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADtwoGiDWDvq9PZ1xQvS9CL8ddGUs2VWks5qYWb-gaJpZM4IJM2C>
> .
>

tomkocse commented:
@vijayaditya Can you point to me the location of your recent rerun of the aspire experiment ?  That helps me for no need to repeat most of the procedures

tomkocse commented:
@vijayaditya I have finished modifying the aspire recipe to use the new  steps/data/reverberate_data_dir.py  . I am verifying if I can repeat your result.

vijayaditya commented:
You need not worry about the replication of results right now as we do not
have spare GPUs. We can test it out later.

Please test the following

   1. Sampling of noise and RIRs is being done appropriately
      1. To test this create non-uniform distributions, sample using your
      script and see if the relative frequencies in the sample are
similar to the
      specified PMF.
   2. Test if the specified noise, SNR and RIR files are being used
   properly by manually listening to the output. Also ensure that the time and
   durations of the point source noises are correct
   3. Ensure there is no mismatch between room-id of isotropic noises
   selected and the RIRs


--Vijay


On Mon, Jul 25, 2016 at 6:40 AM, tomkocse <notifications@github.com> wrote:

> @vijayaditya <https://github.com/vijayaditya> I have finished modifying
> the aspire recipe to use the new steps/data/reverberate_data_dir.py . I am
> verifying if I can repeat your result.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/706#issuecomment-234919980>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADtwoNO-gVbfX4tdcRj8AAmuiX8DeOrKks5qZJKDgaJpZM4IJM2C>
> .
>

vijayaditya commented on egs/aspire/s5/local/multi_condition/aspire_prep_rir_noise_list.py:
```diff
@@ -0,0 +1,79 @@
+    return args
+
+
+# This function generate the rir_list file for the aspire real RIR
```
`This function generates the rir_list file for the real RIRs being in ASpIRE experiments. It assumes the availability of data/impulses_noises directory prepared by local/multi_condition/prepare_impulses_noises.sh.`

vijayaditya commented on egs/aspire/s5/local/multi_condition/aspire_prep_rir_noise_list.py:
```diff
@@ -0,0 +1,79 @@
+#!/usr/bin/env python
```
We will most probably not use this script for long, as we soon plan to create an RIR database and upload it to openslr.org. This script is a stop gap arrangement to use the new `steps/data/reverberate_data_dir.py` until the new upload is prepared. So I am not going ensure that this script is very generic and I will allow for a lot of assumptions to made.

tomkocse replied:
Do you mean you are going to upload the real RIR to openslr.org including the RVB2014, air and RWCP databases?

vijayaditya replied:
I have already prepared around 9 RIR databases for use in our experiments. We have figure out which of these can be redistributed by us and prepare an upload.

vijayaditya commented on egs/aspire/s5/local/multi_condition/aspire_prep_rir_noise_list.py:
```diff
@@ -0,0 +1,79 @@
+
+
+# This function generate the noise_list file from the aspire noise-rir pair 
+def GenerateNoiseListFile(input_dir, output_dir):
```
There is a `data/impulses_noises/info` which has all the information you need.  It defines different rooms and the noises and impulse responses available for that room (e.g. `data/impulses_noises/info/noise_impulse_RVB2014_largeroom1`).

Please do not assign the same `room_id` to all RIRs in a given database. This is not the right usage of `room_id` field.

vijayaditya commented on egs/aspire/s5/local/multi_condition/aspire_prep_rir_noise_list.py:
```diff
@@ -0,0 +1,79 @@
+      type_num = "type*"
+    matched_rir_files = glob.glob(input_dir + "/{0}_{1}_rir_{2}*.wav".format(db_name, type_num, noise_pattern))
+    noise_line = "--noise-id {0} --noise-type isotropic ".format(str(noise_id).zfill(5))
+    for rir in matched_rir_files:
```
Once again use the information in `data/impulses_noises/info/noise_impulse_*` files to get  the noise RIR pairing information.

tomkocse replied:
I have checked that my script can generate the same noise RIR pairing as the original script

vijayaditya commented on egs/aspire/s5/local/multi_condition/run_nnet2_common.sh:
```diff
@@ -7,6 +7,8 @@
 
 stage=1
 snrs="20:10:15:5:0"
```
is `snrs` still necessary ?

tomkocse replied:
not necessary, i will remove it

vijayaditya commented on egs/aspire/s5/local/multi_condition/run_nnet2_common.sh:
```diff
@@ -31,24 +33,29 @@ if [ $stage -le 1 ]; then
     --RIR-home $RIR_home \
     data/impulses_noises || exit 1;
     
+  # Generate the rir_list and noise_list for the reverberate_data_dir.py to corrupt the data
```
# this script just assumes `air rwcp rvb2014` databases.

vijayaditya commented on egs/wsj/s5/steps/data/reverberate_data_dir.py:
```diff
@@ -0,0 +1,506 @@
+    data_lib.RunKaldiCommand("utils/utt2spk_to_spk2utt.pl <{output_dir}/utt2spk >{output_dir}/spk2utt"
+                    .format(output_dir = output_dir))
+
+    if os.path.isfile(input_dir + "/text"):
```
You have to properly prepare an utt2uniq file.

tomkocse replied:
What is the purpose of the utt2uniq file ?

vijayaditya replied:
```bash
 grep -R steps -e utt2uniq
steps/cleanup/combine_short_segments.py:    utt2uniq = None
steps/cleanup/combine_short_segments.py:    if os.path.exists(data_dir_file('utt2uniq')):
steps/cleanup/combine_short_segments.py:        utt2uniq = ParseFileToDict(data_dir_file('utt2uniq'))
steps/cleanup/combine_short_segments.py:    return utt2spk, spk2utt, text, feat, utt2dur, utt2uniq
steps/cleanup/combine_short_segments.py:def WriteCombinedDirFiles(output_dir, utt2spk, spk2utt, text, feat, utt2dur, utt2uniq):
steps/cleanup/combine_short_segments.py:        # updating utt2uniq
steps/cleanup/combine_short_segments.py:        if utt2uniq is not None:
steps/cleanup/combine_short_segments.py:                combined_uniqs.append(utt2uniq.pop(utt))
steps/cleanup/combine_short_segments.py:            # utt2uniq file is used to map perturbed data to original unperturbed
steps/cleanup/combine_short_segments.py:            # is no good way to find the utt2uniq mappinng so that we can avoid
steps/cleanup/combine_short_segments.py:            utt2uniq[new_utt_name] = combined_uniqs[0]
steps/cleanup/combine_short_segments.py:    if utt2uniq is not None:
steps/cleanup/combine_short_segments.py:        WriteDictToFile(utt2uniq, out_dir_file('utt2uniq'))
steps/cleanup/combine_short_segments.py:    utt2spk, spk2utt, text, feat, utt2dur, utt2uniq = ParseDataDirInfo(input_dir)
steps/cleanup/combine_short_segments.py:    WriteCombinedDirFiles(output_dir, utt2spk, spk2utt, text, feat, utt2dur, utt2uniq)
steps/nnet2/get_egs.sh:if [ -f $data/utt2uniq ]; then
steps/nnet2/get_egs.sh:  echo "File $data/utt2uniq exists, so augmenting valid_uttlist to"
steps/nnet2/get_egs.sh:  utils/utt2spk_to_spk2utt.pl $data/utt2uniq > $dir/uniq2utt
steps/nnet2/get_egs.sh:  cat $dir/valid_uttlist.tmp | utils/apply_map.pl $data/utt2uniq | \
steps/nnet2/get_perturbed_feats.sh:# In the combined feature directory, create a file utt2uniq which maps
steps/nnet2/get_perturbed_feats.sh:  > $data/utt2uniq
steps/nnet2/get_egs2.sh:if [ -f $data/utt2uniq ]; then
steps/nnet2/get_egs2.sh:  echo "File $data/utt2uniq exists, so augmenting valid_uttlist to"
steps/nnet2/get_egs2.sh:  utils/utt2spk_to_spk2utt.pl $data/utt2uniq > $dir/uniq2utt
steps/nnet2/get_egs2.sh:  cat $dir/valid_uttlist.tmp | utils/apply_map.pl $data/utt2uniq | \
steps/nnet3/chain/get_egs.sh:if [ -f $data/utt2uniq ]; then  # this matters if you use data augmentation.
steps/nnet3/chain/get_egs.sh:  echo "File $data/utt2uniq exists, so augmenting valid_uttlist to"
steps/nnet3/chain/get_egs.sh:  utils/utt2spk_to_spk2utt.pl $data/utt2uniq > $dir/uniq2utt
steps/nnet3/chain/get_egs.sh:  cat $dir/valid_uttlist.tmp | utils/apply_map.pl $data/utt2uniq | \
steps/nnet3/get_egs_targets.sh:if [ -f $data/utt2uniq ]; then  # this matters if you use data augmentation.
steps/nnet3/get_egs_targets.sh:  echo "File $data/utt2uniq exists, so augmenting valid_uttlist to"
steps/nnet3/get_egs_targets.sh:  utils/utt2spk_to_spk2utt.pl $data/utt2uniq > $dir/uniq2utt
steps/nnet3/get_egs_targets.sh:  cat $dir/valid_uttlist.tmp | utils/apply_map.pl $data/utt2uniq | \
steps/nnet3/get_egs.sh:if [ -f $data/utt2uniq ]; then  # this matters if you use data augmentation.
steps/nnet3/get_egs.sh:  echo "File $data/utt2uniq exists, so augmenting valid_uttlist to"
steps/nnet3/get_egs.sh:  utils/utt2spk_to_spk2utt.pl $data/utt2uniq > $dir/uniq2utt
steps/nnet3/get_egs.sh:  cat $dir/valid_uttlist.tmp | utils/apply_map.pl $data/utt2uniq | \
steps/nnet3/get_egs_discriminative.sh:if [ -f $data/utt2uniq ]; then  # this matters if you use data augmentation.
steps/nnet3/get_egs_discriminative.sh:  echo "File $data/utt2uniq exists, so augmenting valid_uttlist to"
steps/nnet3/get_egs_discriminative.sh:  utils/utt2spk_to_spk2utt.pl $data/utt2uniq > $dir/uniq2utt
steps/nnet3/get_egs_discriminative.sh:  cat $dir/valid_uttlist.tmp | utils/apply_map.pl $data/utt2uniq | \
steps/online/nnet2/get_egs.sh:if [ -f $data/utt2uniq ]; then
steps/online/nnet2/get_egs.sh:  echo "File $data/utt2uniq exists, so augmenting valid/uttlist to"
steps/online/nnet2/get_egs.sh:  utils/utt2spk_to_spk2utt.pl $data/utt2uniq > $dir/uniq2utt
steps/online/nnet2/get_egs.sh:  cat $dir/valid/uttlist.tmp | utils/apply_map.pl $data/utt2uniq | \
steps/online/nnet2/get_egs2.sh:if [ -f $data/utt2uniq ]; then
steps/online/nnet2/get_egs2.sh:  echo "File $data/utt2uniq exists, so augmenting valid/uttlist to"
steps/online/nnet2/get_egs2.sh:  utils/utt2spk_to_spk2utt.pl $data/utt2uniq > $dir/uniq2utt
steps/online/nnet2/get_egs2.sh:  cat $dir/valid/uttlist.tmp | utils/apply_map.pl $data/utt2uniq | \
```

vijayaditya commented:
I agree that your file pattern based pairing might be working. I am suggesting an easier way to generate these pairings as this info is already present in the data/impulses_noises/info file. 


I see a problem in the handling of isotropic noises. Sorry I should noticed this sooner. Isotropic noises are assigned to a room and not an RIR (i.e., room and microphone position). This makes sense as the characteristics of the noise are supposed to be similar in all directions. 

@sw005320 Could you please confirm if the isotropic noise recordings available in RWCP and REVERB-2014 RIR databases can be used with any RIR recorded in the same room.

--Vijay

sw005320 commented:
Yes, RWCP has the isotropic noise recordings.
For example, in http://www.openslr.org/resources/13/RWCP.tar.gz, you can find some noises at RWCP\nospeech\drysrc. 
The REVERB isotropic noises are at the 'NOISE' directories in each-condition directory of  http://reverb2014.dereverberation.com/tools/reverb_tools_for_Generate_SimData.tgz

vijayaditya commented:
@sw005320 Thanks !

vijayaditya commented:
@tomkocse Is it possible to prioritize this PR ? Could we try to get this PR in shape by tomorrow ?

tomkocse commented:
@Vijay I will change the rir linkage of the iso noise to room linkage

vijayaditya commented on egs/aspire/s5/local/multi_condition/aspire_prep_rir_noise_list.py:
```diff
@@ -23,38 +23,48 @@ def GetArgs():
       filename = rir.split('/')[-1]
       if "noise" not in filename:
-        rir_list_file.write('--rir-id {0} --room-id {1} {2}\n'.format(str(rir_id).zfill(5), str(room_id).zfill(3), rir))
+        parts = filename.split('_')
```
Rather than parsing the file name for the necessary parameters, @tomkocse will later submit a modified list generation function which will use information available in `data/impulses_noises/info/`. This is however not a high priority change as we will preprocess the individual databases in the future and these scripts would not be part of any recipe.

vijayaditya commented on egs/aspire/s5/local/multi_condition/aspire_prep_rir_noise_list.py:
```diff
@@ -23,38 +23,48 @@ def GetArgs():
-        rir_list_file.write('--rir-id {0} --room-id {1} {2}\n'.format(str(rir_id).zfill(5), str(room_id).zfill(3), rir))
+        parts = filename.split('_')
+        db_name = parts[0]
+        type_num = parts[1]
```
I am assuming these filename parsing code is correct.

vijayaditya commented on src/featbin/wav-reverberate.cc:
```diff
@@ -133,22 +176,37 @@ int main(int argc, char *argv[]) {
+                "If nonzero, it specified the duration (secs) of the output "
+                "signal. If the duration t is less than the length of the "
+                "input signal, the first t secs of the signal is trimed, "
+                "otherwise, the signal will be repeated to"
```
specify that this option is used for extending the length of isotropic noises.

vijayaditya commented:
There are minor changes necessary in this PR, however these are not high priority. @tomkocse Would continue work on those changes in a different PR. I would like to merge these scripts now as this a high priority requirement for some of the experiments we are working on. These scripts are isolated from the rest of the code base and are not used in any checked-in recipe other than ASpIRE.

@danpovey , @jtrmal could you please perform a second review and perform the merge once @tomkocse resolves the conflicts.

tomkocse commented:
Please go to #927 for the continuation of this PR
